appserver/webservices/jsr109-impl/src/main/java/org/glassfish/webservices/monitoring/WebServiceTesterServlet.java

```diff
@@ -646,4 +653,64 @@ private void deleteDir(File path) {
     private String encodeHTML(String html) {
         return html.replaceAll("<", "&lt;").replaceAll(">", "&gt;");
     }
+
+    /**
+     * Method to test that the Host Header is valid and doesn't have any illegal characters, to help prevent Host Header
+     * Attacks.
+     *
+     * @param request The {@link HttpServletRequest} from the
+     * {@link WebServiceTesterServlet#invoke(HttpServletRequest, HttpServletResponse, WebServiceEndpoint)} method
+     * @return True if the Host Header is valid, false if it isn't
+     */
+    private boolean isValidHostHeader(HttpServletRequest request) {
+        // Get the Host Header
+        String serverName = request.getServerName();
+
+        // Decode the Host Header
+        String decodedServerName = null;
+        try {
+            decodedServerName = URLDecoder.decode(serverName, "UTF-8");
+        } catch (UnsupportedEncodingException unsupportedEncodingException) {
+            logger.log(Level.WARNING, "Could not decode URL during validation, assuming Host Header is invalid.",
+                    unsupportedEncodingException);
+            return false;
+        }
+
+        // Shouldn't Happen™, but just in case
+        if (decodedServerName == null) {
+            logger.log(Level.WARNING, "Host Header decoded to null during validation, assuming Host Header is invalid.");
+            return false;
+        }
+
+        // Check IPv6 literal: []
+        if (decodedServerName.startsWith("[") && decodedServerName.endsWith("]")) {
+            return true;
+        }
+
+        // Check IPv4 dotted-decimal: xxx.xxx.xxx.xxx
+        if (checkValidIpv4DottedDecimal(decodedServerName)) {
+            return true;
+        }
+
+        // Check DNS name is valid - ASCII alphanumerics and "-" only, no segment longer than 63 characters,
+        // sections separated with ".", possibly ending with "."
+        if (checkValidDnsName(decodedServerName)) {
+            return true;
+        }
+
+        // No match implies invalid host header / server name
+        return false;
+    }
+
+    // Split off for unit testing
+    protected static boolean checkValidIpv4DottedDecimal(String decodedServerName) {
+        return decodedServerName.matches(
+                "^((0|1\\d?\\d?|2[0-4]?\\d?|25[0-5]?|[3-9]\\d?)\\.){3}(0|1\\d?\\d?|2[0-4]?\\d?|25[0-5]?|[3-9]\\d?)$");
+    }
+
+    // Split off for unit testing
+    protected static boolean checkValidDnsName(String decodedServerName) {
+        // NOTE: HttpServletRequest.getServerName strips the port so no need to check
+        return decodedServerName.matches("((?!-)[\\w-]{1,63}(?<!-)(\\.|$))+");
```

Comment:
I think that rules depend on some local authority. What it will do with IDN?
https://en.wikipedia.org/wiki/Internationalized_domain_name
https://en.wikipedia.org/wiki/Punycode
https://docs.oracle.com/javase/8/docs/api/java/net/IDN.html
I am not sure if the UrlDecoder is the right one for this ...

EDIT: HTTP request to an url using my custom hostname řízeček:8080 passed, so at least this is an example of an ugly header. Translation was done by the browser, but I had to use the translated value in /etc/hosts
![image](https://user-images.githubusercontent.com/302904/115129511-f2384300-9fe6-11eb-9765-bbc21ba72e63.png)

Reply:
UrlDecoder is used elsewhere in this code, which is where I took it from.
Patrik and myself looked at IDNs - RFC for DNS states something along the lines (can't remember exactly) of TLDR: ASCII characters only.